### PR TITLE
enable auto redeploy for hidden model

### PR DESCRIPTION
**Pull request by rbhavna**
### Description
enable auto redeploy for hidden model
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: common/src/main/java/org/opensearch/ml/common/transport/deploy/MLDeployModelRequest.java [hunk 1]
```diff
@@ -38,24 +38,29 @@ public class MLDeployModelRequest extends MLTaskRequest {
...
         super(dispatchTask);
         this.modelId = modelId;
         this.modelNodeIds = modelNodeIds;
         this.async = async;
+        this.isUserInitiatedDeployRequest = isUserInitiatedDeployRequest;
     }
 
     public MLDeployModelRequest(String modelId, boolean async) {
-        this(modelId, null, async, true);
+        this(modelId, null, async, true, true);
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/transport/deploy/MLDeployModelRequest.java:55]: [nit] may be add a comment why here `isUserInitiatedDeployRequest` will always be true?

File: common/src/main/java/org/opensearch/ml/common/transport/deploy/MLDeployModelRequest.java [hunk 2]
```diff
@@ -38,24 +38,29 @@ public class MLDeployModelRequest extends MLTaskRequest {
...
     }
 
     public MLDeployModelRequest(StreamInput in) throws IOException {
         super(in);
         this.modelId = in.readString();
         this.modelNodeIds = in.readOptionalStringArray();
         this.async = in.readBoolean();
+        this.isUserInitiatedDeployRequest = in.readOptionalBoolean();
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/transport/deploy/MLDeployModelRequest.java:63]: why optionalboolean? what'll happen if the value isn't present?

File: plugin/src/test/java/org/opensearch/ml/action/deploy/TransportDeployModelActionTests.java [hunk 3]
```diff
@@ -286,6 +286,7 @@ public void testDoExecute_no_permission_hidden_model() {
         MLModel mlModel = mock(MLModel.class);
         when(mlModel.getAlgorithm()).thenReturn(FunctionName.ANOMALY_LOCALIZATION);
         when(mlModel.getIsHidden()).thenReturn(true);
+        when(mlDeployModelRequest.isUserInitiatedDeployRequest()).thenReturn(true);
```
Comment (arjunkumargiri) [inline plugin/src/test/java/org/opensearch/ml/action/deploy/TransportDeployModelActionTests.java:289]: Add unit test to cover `isUserInitiatedDeployRequest` set to false.

**Review by arjunkumargiri — COMMENTED** [review]
Can we follow similar logic for non-hidden models to bypass access check in case the request is system initiated? : https://github.com/opensearch-project/ml-commons/pull/2102/files#diff-c341f5a67c1d221e5593dd55fb689e9d0c9a60d33e0fa26090803cfa90818fa7R162

**Comment by rbhavna** [comment]
> Can we follow similar logic for non-hidden models to bypass access check in case the request is system initiated? : https://github.com/opensearch-project/ml-commons/pull/2102/files#diff-c341f5a67c1d221e5593dd55fb689e9d0c9a60d33e0fa26090803cfa90818fa7R162

This is already handled in the validateModelGroupAccess method. But it checks for user being null and skips access check which handles the auto redeploy scenario

File: plugin/src/main/java/org/opensearch/ml/model/MLModelManager.java [hunk 4]
```diff
@@ -857,7 +857,7 @@ private void updateModelRegisterStateAsDone(
     void deployModelAfterRegistering(MLRegisterModelInput registerModelInput, String modelId) {
         String[] modelNodeIds = registerModelInput.getModelNodeIds();
         log.debug("start deploying model after registering, modelId: {} on nodes: {}", modelId, Arrays.toString(modelNodeIds));
-        MLDeployModelRequest request = new MLDeployModelRequest(modelId, modelNodeIds, false, true);
+        MLDeployModelRequest request = new MLDeployModelRequest(modelId, modelNodeIds, false, true, true);
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/model/MLModelManager.java:860]: What if user define a flow template with deploy hidden model action in it ? Can user deploy any hidden model?

Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/model/MLModelManager.java:860]: Isn't hidden model management (register, deploy, undeploy, delete) is out of scope for AI-Flow?

**Comment by arjunkumargiri** [comment]
> > Can we follow similar logic for non-hidden models to bypass access check in case the request is system initiated? : https://github.com/opensearch-project/ml-commons/pull/2102/files#diff-c341f5a67c1d221e5593dd55fb689e9d0c9a60d33e0fa26090803cfa90818fa7R162
> 
> This is already handled in the validateModelGroupAccess method. But it checks for user being null and skips access check which handles the auto redeploy scenario

Null user checks is a workaround, can we explicitly check if deploy request is not user initiated so that same behavior is maintained for system initiated deployments?

File: plugin/src/main/java/org/opensearch/ml/model/MLModelManager.java (hunk 4 above)
Comment (rbhavna) [reply plugin/src/main/java/org/opensearch/ml/model/MLModelManager.java:860]: With a flow template, we will still consider it as user initiated right. So we will still check for access. Only if it is system initiated, we will bypass the access checks

Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/model/MLModelManager.java:860]: > Isn't hidden model management (register, deploy, undeploy, delete) is out of scope for AI-Flow?

Yes for normal process, but that's not a hard control. What if user just copy the hidden model id and define a flow template to deploy it?

**Review by arjunkumargiri — APPROVED** [review]

File: plugin/src/main/java/org/opensearch/ml/model/MLModelManager.java (hunk 4 above)
Comment (zane-neo) [reply plugin/src/main/java/org/opensearch/ml/model/MLModelManager.java:860]: Can user see the hidden model id? I thought only superAdmin can see it.

Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/model/MLModelManager.java:860]: >Can user see the hidden model id? I thought only superAdmin can see it.

model id isn't hidden to users. But model content is hidden to users. During prediction, model id is being used. As user can make the prediction request, they can know the model id eventually.

Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/model/MLModelManager.java:860]: >Yes for normal process, but that's not a hard control. User can copy the hidden model id and define a flow template to deploy it.

We opened this [deploy request ](https://github.com/opensearch-project/ml-commons/blob/main/client/src/main/java/org/opensearch/ml/client/MachineLearningNodeClient.java#L260) to clients and then user initiated request flag is being set to true from code. So every request sent from ai-flow or any other clients should be treated as user initiated request too.

File: common/src/main/java/org/opensearch/ml/common/transport/deploy/MLDeployModelRequest.java [hunk 5]
```diff
@@ -38,24 +38,32 @@ public class MLDeployModelRequest extends MLTaskRequest {
     private String modelId;
     private String[] modelNodeIds;
     boolean async;
+    // This is to identify if the get request is initiated by user or not. During auto redeploy also, we perform deploy operation.
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/transport/deploy/MLDeployModelRequest.java:41]: [nit] this is a deploy request not a get request.

**Comment by dhrubo-os** [comment]
> > > Can we follow similar logic for non-hidden models to bypass access check in case the request is system initiated? : https://github.com/opensearch-project/ml-commons/pull/2102/files#diff-c341f5a67c1d221e5593dd55fb689e9d0c9a60d33e0fa26090803cfa90818fa7R162
> > 
> > 
> > This is already handled in the validateModelGroupAccess method. But it checks for user being null and skips access check which handles the auto redeploy scenario
> 
> Null user checks is a workaround, can we explicitly check if deploy request is not user initiated so that same behavior is maintained for system initiated deployments?

@rbhavna are we planning to address this?

**Comment by rbhavna** [comment]
> > > > Can we follow similar logic for non-hidden models to bypass access check in case the request is system initiated? : https://github.com/opensearch-project/ml-commons/pull/2102/files#diff-c341f5a67c1d221e5593dd55fb689e9d0c9a60d33e0fa26090803cfa90818fa7R162
> > > 
> > > 
> > > This is already handled in the validateModelGroupAccess method. But it checks for user being null and skips access check which handles the auto redeploy scenario
> > 
> > 
> > Null user checks is a workaround, can we explicitly check if deploy request is not user initiated so that same behavior is maintained for system initiated deployments?
> 
> @rbhavna are we planning to address this?

Yes this has been addressed already https://github.com/opensearch-project/ml-commons/pull/2102/files#diff-c341f5a67c1d221e5593dd55fb689e9d0c9a60d33e0fa26090803cfa90818fa7R147

**Review by dhrubo-os — APPROVED** [review]

**Review by ylwu-amzn — APPROVED** [review]

**Comment by Zhangxunmt** [comment]
Why this PR is not backported to main?

**Comment by rbhavna** [comment]
> Why this PR is not backported to main?

This was merged in this PR @Zhangxunmt https://github.com/opensearch-project/ml-commons/pull/2136

**Pull request merged**